dvc/repo/artifacts.py

```diff
@@ -0,0 +1,91 @@
+import os
+from typing import TYPE_CHECKING
+
+from dvc.annotations import Artifact
+from dvc.dvcfile import FileMixin, ProjectFile
+
+if TYPE_CHECKING:
+    from dvc.repo import Repo
+
+
+class ArtifactsFile(FileMixin):
+    from dvc.schema import ARTIFACTS_SCHEMA as SCHEMA
+
+
+class Artifacts:
+    repo: "Repo"
+
+    def __init__(self, repo) -> None:
+        self.repo = repo
+
+    def _read(self):
+        # merge artifacts from all dvc.yaml files found
+        artifacts = {}
+        for dvcfile, dvcfile_artifacts in self.repo.index._artifacts.items():
+            # read the artifacts.yaml file if needed
+            if isinstance(dvcfile_artifacts, str):
+                dvcfile_artifacts = ArtifactsFile(
+                    self.repo,
+                    os.path.join(os.path.dirname(dvcfile), dvcfile_artifacts),
+                    verify=False,
+                ).load()
+            for name, value in dvcfile_artifacts.items():
+                if name in artifacts:
+                    # maybe better to issue a warning here and take the first one?
+                    # QUESTION: how to deal with the same artifact name in different dvc.yaml files?
+                    raise ValueError(
+                        f"Duplicated artifact ID: {name} in {dvcfile} and {artifacts[name]['dvcfile']}"
+                    )
+                artifacts[name] = {"dvcfile": dvcfile, "annotation": Artifact(**value)}
+        return artifacts
+
+    def read(self):
+        return {name: value["annotation"] for name, value in self._read().items()}
+
+    def get(self, name):
+        return self.read().get(name, Artifact())
+
+    # TODO: big question: how this will work for DVC subprojects?
+    # i.e. how should we match Git tags with subprojects, if the same artifacts exists in both?
+    # e.g. how Studio should figure out what Git tags belong to this subproject?
+    # e.g. how you figure out what artifact to take in CI/CD?
+    def _resolve_dvcfile_path(self, name=None, must_exist=False, prefered_path=None):
+        artifacts = self._read()
+        # if artifact with given name already exists, give the path to its file
+        if name:
+            if name in artifacts:
+                return artifacts[name]["dvcfile"]
+            if must_exist:
+                raise ValueError(f"No artifact {name} found")
+        # if `prefered_path` is given, use it
+        # this will allow to write to dvclive's dvc.yaml
+        # QUESTION 1: is it OK that dvclive.log_artifact() will update existing artifact that's in some "external" dvc.yaml?
```

Comment:
@dberenbaum WDYT? I think this is how it should work.

Reply:
I don't think we need to have `add` or `remove` methods in DVC right now, which should hopefully minimize a lot of these questions. Let's leave it up to the users to write this manually or use `Live.log_artifact()`. That's similar to the support we have for top-level plots so far.

Reply:
I don't mean exposing API/CLI for users, I mean exposing API for Studio and DVCLive. Both Studio and `live.log_artifacts` will `add` artifacts, so it make sense to implement it in DVC I believe. Otherwise we're getting code duplication, potential breakages if we update something here, and logic we don't control (maybe giving DVCLive/Studio to control this logic is actually good - they can implement it their way, but so far I don't think so). WDYT?

The question I wanted to ask here ` is it OK that dvclive.log_artifact(myart) will update existing `myart` artifact that  exists in some "external" dvc.yaml (not in dvclive/dvc.yaml)?` I guess it's ok, but you raised a concern about DVCLive writing to such files - and I don't know what's the problem with it. cc @daavoo

Reply:
> is it OK that dvclive.log_artifact(myart) will update existingmyart artifact that  exists in some "external" dvc.yaml (not in dvclive/dvc.yaml)? I guess it's ok, but you raised a concern about DVCLive writing to such files - and I don't know what's the problem with it. cc @daavoo

I am taking a step back and trying to check if we are all on the same page.

The way I understand the `artifacts` section is similar to the current `metrics`/`params``plots` top-level sections, which are completely descriptive.

We start from the "DVCLive only" onboarding scenario.

`Live.log_artifact` currently does one thing, which is to update the `.dvc` file (creating it if it's needed).

The only new action that would be needed is to write the `artifacts` section in the `dvclive/dvc.yaml`, similar to how we currently write `metrics`,`params`,`plots`. Supporting it should be a simple update to the `yaml.dump`.

It has nothing to do with `log_artifact` or other `log_{X}` calls. 
It is done by [`make_dvcyaml`](https://dvc.org/doc/dvclive/live/make_dvcyaml) and is handled by a `[dvcyaml`](https://dvc.org/doc/dvclive/live#parameters) arg.

---

Any other scenario (i.e. user now migrates to pipelines and adds a root`dvc.yaml`, monorepo, etc) we just assume that the user will either rely on the descriptions already present in `dvclive/dvc.yaml` or would deactivate them (`Live(dvcyaml=False)`) in favor of manually written them in other `dvc.yaml`.

I feel like I might be missing where the complexity we are currently discussing comes from

Reply:
Ok, I didn't know that. Thanks @daavoo. Then I guess the right way to work with this would be:
1. If DVCLive was set up to write to `dvc.yaml`, use DVC API to write to `{Live.dir}/dvc.yaml` when `log_artifact` is called.
2. At this step, if `myartifactname` exists, DVC will throw an error since no `name`s duplication is allowed. How we handle this?
3. if DVCLive wasn't set up to write to `dvc.yaml` the way you described, but user calls `log_artifact(path, name="myname")` or pass any other args (like `labels`, `type` or `description`), what should DVCLive do? I guess the answer should be similar to the case when one calls `log_metric`, but forgets to call `make_dvcyaml` at the end. What happens then, could you please explain?

Reply:
ok, so actually @dberenbaum posted this https://github.com/iterative/dvclive/issues/472#issuecomment-1470765436 which is basically about the same.

Reply:
ok, so actually @dberenbaum posted this https://github.com/iterative/dvclive/issues/472#issuecomment-1470765436 which is basically about the same.

Reply:
@aguschin Let's keep it simple for now. When reading, we should check for duplicates because we have to do something, but when writing, let's not worry about it yet.

> Studio and DVCLive

For DVCLive, I think it's enough to have DVCLive act like a DVC user and write to `dvclive/dvc.yaml` manually like it does for plots.

Studio should be able to do something similar even if it means some duplication (should be pretty minor if it's just writing to yaml files).